runtime/vm/stringhelpers.cpp

```diff
@@ -258,131 +258,67 @@ compareStringToUTF8(J9VMThread *vmThread, j9object_t string, UDATA translateDots
 
-	if (compressed) {
-		/* following implementation is expected to be different from non-compressed case when String compressing is enabled */
-		for (i = unicodeOffset; i < unicodeOffset + unicodeLength; i++) {
-			result = VM_VMHelpers::encodeUTF8CharN(J9JAVAARRAYOFBYTE_LOAD(vmThread, unicodeBytes, i), data, (U_32)remaining);
-			if (0 == result) {
-				return UDATA_MAX;
-			} else {
-				if ((stringFlags & J9_STR_XLAT) && ('.' == *data)) {
+	if (IS_STRING_COMPRESSED(vmThread, string)) {
+		for (UDATA i = 0; i < unicodeLength; i++) {
+			data += VM_VMHelpers::encodeUTF8CharI8(J9JAVAARRAYOFBYTE_LOAD(vmThread, unicodeBytes, i), data);
+		}
+
+		if (stringFlags & J9_STR_XLAT) {
+			data = utf8Data;
+
+			for (UDATA i = 0; i < unicodeLength; i++) {
+				UDATA encodedLength = VM_VMHelpers::encodedUTF8LengthI8(J9JAVAARRAYOFBYTE_LOAD(vmThread, unicodeBytes, i));
+
+				if ('.' == *data) {
 					*data = '/';
 				}
-				remaining -= result;
-				data += result;
+
+				data += encodedLength;
 			}
 		}
 	} else {
-		for (i = unicodeOffset; i < (unicodeOffset + unicodeLength); i++) {
-			result = VM_VMHelpers::encodeUTF8CharN(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i), data, (U_32)remaining);
-			if (0 == result) {
-				return UDATA_MAX;
-			} else {
-				if ((stringFlags & J9_STR_XLAT) && ('.' == *data)) {
-					*data = '/';
-				}
-				remaining -= result;
-				data += result;
-			}
+		for (UDATA i = 0; i < unicodeLength; i++) {
+			data += VM_VMHelpers::encodeUTF8Char(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i), data);
 		}
-	}
-	if (nullTermination && (0 == remaining)) {
-		return UDATA_MAX;
-	}
-	if (nullTermination) {
-		*data = '\0';
-	}
 
-	return data - utf8Data;
-}
+		if (stringFlags & J9_STR_XLAT) {
+			data = utf8Data;
 
-/**
- * Copy a Unicode String to a UTF8 data buffer with NULL termination.
- *
- * @param[in] vmThread the current J9VMThread
- * @param[in] string a string object to be copied, it can't be NULL
- * @param[in] nullTermination if the utf8 data is going to be NULL terminated
- * @param[in] stringFlags the flag to determine performing '.' --> '/'
- * @param[in] utf8Data a utf8 data buffer
- * @param[in] utf8Length the size of the utf8 data buffer
- *
- * @return UDATA_MAX if a failure occurred, otherwise the number of utf8 data copied excluding null termination
- */
-UDATA
-copyStringToUTF8Helper(J9VMThread *vmThread, j9object_t string, BOOLEAN nullTermination, UDATA stringFlags, U_8 *utf8Data, UDATA utf8Length)
-{
-	Assert_VM_notNull(string);
-
-	UDATA unicodeLength = J9VMJAVALANGSTRING_LENGTH(vmThread, string);
-	j9object_t unicodeBytes = J9VMJAVALANGSTRING_VALUE(vmThread, string);
-
-	return copyCharsIntoUTF8Helper(vmThread, IS_STRING_COMPRESSED(vmThread, string), nullTermination, stringFlags, unicodeBytes, 0, unicodeLength, utf8Data, utf8Length);
-}
+			for (UDATA i = 0; i < unicodeLength; i++) {
+				UDATA encodedLength = VM_VMHelpers::encodedUTF8Length(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i));
```

Comment:
Can `stringFlags` check along with potential `. --> /` replacement be performed within last `for` loop to avoid duplicated call of `J9JAVAARRAYOFCHAR_LOAD()`?

Reply:
I'm not sure what you mean by this. The `stringFlags` check is performed before the last for loop. We cannot avoid `J9JAVAARRAYOFCHAR_LOAD` here because we only replace the first byte of each individual code point, so we need to traverse the array by code points not by bytes.

Reply:
Current code:
````
		for (i = unicodeOffset; i < (unicodeOffset + unicodeLength); i++) {
			result = VM_VMHelpers::encodeUTF8CharN(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i), data, (U_32)remaining);
			if (0 == result) {
				return UDATA_MAX;
			} else {
				if ((stringFlags & J9_STR_XLAT) && ('.' == *data)) {
					*data = '/';
				}
				remaining -= result;
				data += result;
			}
		}
````

This Pull Request:
````
		for (UDATA i = 0; i < unicodeLength; i++) {
			data += VM_VMHelpers::encodeUTF8Char(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i), data);
		}

		if (stringFlags & J9_STR_XLAT) {
			data = utf8Data;

			for (UDATA i = 0; i < unicodeLength; i++) {
				UDATA encodedLength = VM_VMHelpers::encodedUTF8Length(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i));

				if ('.' == *data) {
					*data = '/';
				}

				data += encodedLength;
			}
		}
````
Why introduce an extra loop (and duplicated code) to handle `J9_STR_XLAT` which might hurt startup performance? 
Does it intend to avoid comparison of `J9_STR_XLAT` inside of first loop and improve the application runtime?
If so, I would suggest to make two version of `copyCharsIntoUTF8HelperXXX`, one for `J9_STR_XLAT` check during startup, the other without `J9_STR_XLAT` for application runtime, and they could share a helper invoking `VM_VMHelpers::encodeUTF8Char`.
Does it make sense?

Reply:
As per my performance evaluation comment (https://github.com/eclipse/openj9/pull/1047#issuecomment-361365752)

> The compiler does not version these outside the loop. The main beneficial change here is to manually version these out of the tight for loop to allow for a very fast sequence.

I dislike the idea of introducing another `copyCharsIntoUTF8HelperXXX` just to handle the case of `J9_STR_XLAT`. If we go down that route should we not specialize every single function which takes in any sort of flag since we mostly pass flags as constants? At what point do you make the cutoff then?

How about this best of both worlds:

```
if ((stringFlags & J9_STR_XLAT) == 0) {
    for (UDATA i = 0; i < unicodeLength; i++) {
        data += VM_VMHelpers::encodeUTF8Char(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i), data);
    }
} else {
    for (UDATA i = 0; i < unicodeLength; i++) {
        UDATA encodedLength = VM_VMHelpers::encodedUTF8Length(J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i));

        if ('.' == *data) {
            *data = '/';
        }

        data += encodedLength;
    }
}
```

This eliminates the extra `J9JAVAARRAYOFCHAR_LOAD` in the second for loop and we still perform fast in both cases. Thoughts?

Reply:
This works. Some comments about this arrangement will ensure that no _optimization_ to remove this optimization.

Reply:
Fixed in bced3dc